File: lib/core/services/announcement_service.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:hilla_ride/core/models/announcement.dart';
import 'package:shared_preferences/shared_preferences.dart';

class AnnouncementService {
  AnnouncementService({FirebaseFirestore? firestore})
      : _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _firestore;
  final _readIdsController = StreamController<Set<String>>.broadcast();
  static const _readIdsKey = 'read_announcement_ids';

  Stream<List<Announcement>> watchAnnouncements(String audience) {
    return _firestore
        .collection('announcements')
        .where('audience', isEqualTo: audience)
        .limit(40)
        .snapshots()
        .map((snapshot) {
      final items = snapshot.docs
          .map((doc) => Announcement.fromMap(doc.id, doc.data()))
          .toList();
      items.sort((a, b) {
        final aTime = a.createdAt ?? DateTime.fromMillisecondsSinceEpoch(0);
        final bTime = b.createdAt ?? DateTime.fromMillisecondsSinceEpoch(0);
        return bTime.compareTo(aTime);
      });
      return items;
    });
  }

  Stream<Set<String>> watchReadIds() async* {
    yield await getReadIds();
    yield* _readIdsController.stream;
  }

  Future<Set<String>> getReadIds() async {
    final prefs = await SharedPreferences.getInstance();
    final list = prefs.getStringList(_readIdsKey) ?? const [];
    return list.toSet();
  }

  Future<void> _publishReadIds() async {
    if (_readIdsController.isClosed) return;
    _readIdsController.add(await getReadIds());
  }

  Future<void> markRead(String id) async {
    final prefs = await SharedPreferences.getInstance();
    final read = await getReadIds();
    if (read.contains(id)) return;
    read.add(id);
    await prefs.setStringList(_readIdsKey, read.toList());
    await _publishReadIds();
  }

  Future<void> markAllRead(Iterable<String> ids) async {
    final prefs = await SharedPreferences.getInstance();
    final read = await getReadIds();
    read.addAll(ids);
    await prefs.setStringList(_readIdsKey, read.toList());
    await _publishReadIds();
  }

  int unreadCount(List<Announcement> announcements, Set<String> readIds) {
    return announcements.where((item) => !readIds.contains(item.id)).length;
  }
}

File: hilla_ride_ios/HelloTukTuk/Core/Services/AnnouncementService.swift
import FirebaseFirestore
import Foundation

final class AnnouncementService {
    private let firestore = Firestore.firestore()
    private let readIdsKey = "read_announcement_ids"

    private func dismissedBannerKey(audience: String) -> String {
        "dismissed_banner_ids_\(audience)"
    }

    func watchAnnouncements(audience: String) -> AsyncStream<[Announcement]> {
        AsyncStream { continuation in
            let listener = firestore.collection("announcements")
                .whereField("audience", isEqualTo: audience)
                .limit(to: 40)
                .addSnapshotListener { snapshot, _ in
                    let items = snapshot?.documents.compactMap { doc in
                        Announcement(documentID: doc.documentID, data: doc.data())
                    } ?? []
                    let sorted = items.sorted {
                        ($0.createdAt ?? .distantPast) > ($1.createdAt ?? .distantPast)
                    }
                    continuation.yield(sorted)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func watchActiveBanners(audience: String) -> AsyncStream<[Announcement]> {
        AsyncStream { continuation in
            let listener = firestore.collection("announcements")
                .whereField("audience", isEqualTo: audience)
                .limit(to: 20)
                .addSnapshotListener { snapshot, _ in
                    let items = snapshot?.documents.compactMap { doc in
                        Announcement(documentID: doc.documentID, data: doc.data())
                    } ?? []
                    let sorted = items.sorted {
                        ($0.createdAt ?? .distantPast) > ($1.createdAt ?? .distantPast)
                    }
                    let bannerItems = Array(sorted.filter(\.showAsBanner).prefix(5))
                    continuation.yield(bannerItems.isEmpty ? Array(sorted.prefix(5)) : bannerItems)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func getDismissedBannerIds(audience: String) -> Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: dismissedBannerKey(audience: audience)) ?? [])
    }

    func dismissBanner(id: String, audience: String) {
        var dismissed = getDismissedBannerIds(audience: audience)
        guard !dismissed.contains(id) else { return }
        dismissed.insert(id)
        UserDefaults.standard.set(Array(dismissed), forKey: dismissedBannerKey(audience: audience))
    }

    func getReadIds() -> Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: readIdsKey) ?? [])
    }

    func markRead(_ id: String) {
        var read = getReadIds()
        guard !read.contains(id) else { return }
        read.insert(id)
        UserDefaults.standard.set(Array(read), forKey: readIdsKey)
    }

    func markAllRead(_ ids: [String]) {
        var read = getReadIds()
        ids.forEach { read.insert($0) }
        UserDefaults.standard.set(Array(read), forKey: readIdsKey)
    }

    func unreadCount(announcements: [Announcement], readIds: Set<String>) -> Int {
        announcements.filter { !readIds.contains($0.id) }.count
    }
}
